File: cmd/claude.go
package cmd

import (
	"bufio"
	"crypto/rand"
	"fmt"
	"math/big"
	"os"
	"os/exec"
	"strings"
	"syscall"

	"github.com/mturley/agent-handler/terminal"
	"github.com/spf13/cobra"
)

var claudeCmd = &cobra.Command{
	Use:                "claude [claude-args...]",
	Short:              "Start a peekable Claude session",
	Long:               "Wrapper that ensures the Claude session is peekable via handler peek, then passes all arguments through to claude.",
	DisableFlagParsing: true,
	RunE:               runClaude,
}

func init() {
	claudeCmd.GroupID = "human"
	rootCmd.AddCommand(claudeCmd)
}

func runClaude(cmd *cobra.Command, args []string) error {
	claudeBin, err := exec.LookPath("claude")
	if err != nil {
		return fmt.Errorf("claude not found on PATH: %w", err)
	}

	backendType, _ := terminal.Detect()

	switch backendType {
	case "cmux":
		os.Setenv("HANDLER_MANAGED", "1")
		return syscall.Exec(claudeBin, append([]string{"claude"}, args...), os.Environ())

	case "tmux":
		// Set pane title to handler:pending
		exec.Command("tmux", "select-pane", "-T", "handler:pending").Run()
		os.Setenv("HANDLER_MANAGED", "1")
		return syscall.Exec(claudeBin, append([]string{"claude"}, args...), os.Environ())

	default:
		// Outside both — prompt user
		fmt.Println("No tmux or cmux detected. Start a tmux session for peek support? [y/N]")
		reader := bufio.NewReader(os.Stdin)
		answer, _ := reader.ReadString('\n')
		answer = strings.TrimSpace(strings.ToLower(answer))

		if answer == "y" || answer == "yes" {
			suffix, _ := rand.Int(rand.Reader, big.NewInt(99999))
			sessionName := fmt.Sprintf("handler-%05d", suffix.Int64())
			// Build the claude command string for tmux
			claudeArgs := strings.Join(args, " ")
			claudeCommand := claudeBin
			if claudeArgs != "" {
				claudeCommand = claudeBin + " " + claudeArgs
			}
			tmuxCmd := exec.Command("tmux", "new-session", "-s", sessionName,
				"-e", "HANDLER_MANAGED=1",
				claudeCommand)
			tmuxCmd.Stdin = os.Stdin
			tmuxCmd.Stdout = os.Stdout
			tmuxCmd.Stderr = os.Stderr
			// Set the pane title after session creation
			exec.Command("tmux", "select-pane", "-t", sessionName, "-T", "handler:pending").Run()
			return tmuxCmd.Run()
		}

		// User declined — run without peek support
		return syscall.Exec(claudeBin, append([]string{"claude"}, args...), os.Environ())
	}
}

File: src/Handler/Cmd/Claude.hs
-- | Port of cmd/claude.go: start a peekable Claude session.
module Handler.Cmd.Claude (claudeCommand) where

import Control.Monad (void)
import Data.Text (Text)
import qualified Data.Text as T
import Options.Applicative
import System.Directory (findExecutable)
import System.Environment (setEnv)
import System.Exit (exitWith)
import System.IO (hFlush, stdout)
import System.Posix.Process (executeFile)
import System.Process (readProcessWithExitCode, spawnProcess, waitForProcess)
import System.Random (randomRIO)
import Text.Printf (printf)

import Handler.Cli.Common (Ctx, NamedCommand, dieText, putTextLn)
import Handler.Terminal (detect)

-- | Flag parsing is disabled (cobra DisableFlagParsing): everything after
-- \"claude\" is forwarded verbatim, so no helper wrapper here.
claudeCommand :: Mod CommandFields NamedCommand
claudeCommand =
  command "claude" $ info
    (fmap ((,) "claude") (runClaude <$> many (strArgument (metavar "CLAUDE-ARGS"))))
    ( progDesc "Start a peekable Claude session"
    <> forwardOptions <> noIntersperse )

runClaude :: [Text] -> Ctx -> IO ()
runClaude args _ = do
  mclaudeBin <- findExecutable "claude"
  case mclaudeBin of
    Nothing -> dieText "claude not found on PATH"
    Just claudeBin -> do
      (backendType, _, _) <- detect
      let execClaude = do
            setEnv "HANDLER_MANAGED" "1"
            executeFile claudeBin False (map T.unpack args) Nothing
      case backendType of
        "cmux" -> execClaude
        "tmux" -> do
          -- Set pane title to handler:pending
          void $ readProcessWithExitCode "tmux" ["select-pane", "-T", "handler:pending"] ""
          execClaude
        _ -> do
          -- Outside both — prompt user
          putTextLn "No tmux or cmux detected. Start a tmux session for peek support? [y/N]"
          hFlush stdout
          answer <- T.toLower . T.strip . T.pack <$> getLine
          if answer == "y" || answer == "yes"
            then do
              suffix <- randomRIO (0, 99999 :: Int)
              let sessionName = printf "handler-%05d" suffix :: String
              ph <- spawnProcess "tmux"
                ([ "new-session", "-s", sessionName
                 , "-e", "HANDLER_MANAGED=1", claudeBin
                 ] ++ map T.unpack args)
              code <- waitForProcess ph
              exitWith code
            else execClaude
